### ISSUE TITLE
[BUG] Button messages are not visible on the latest version in android devices

### ISSUE THREAD

**Issue by TentacleSama4254** [adiwajshing/Baileys#2631]
**To Reproduce**
1. Created a new connection
2. Closed & used saved credentials to log back in
3. sens a ButtonsMessage to a Android device whith the latest version of Whtapp installed

**Expected behavior**
Buttons to be visible and clickable

**Expected behavior**
- Buttons visible to sender but not visible to receiver id they are on android
- old buttons messages received are still visible
- visible to ios devices

**Environment (please complete the following information):**
 - tested both on a server and locally
 - tested sending with both ios and android devices
 - tested with both Whatsapp and Whatsapp Business

**Comment by RasyidMF** [comment]
Someone ever answer this problem and i solved by his solutions.. add the property key that name as ```patchMessageBeforeSending``` in ```makeWASocket``` function and add this code

```js
makeWASocket({
patchMessageBeforeSending: (message) => {
                const requiresPatch = !!(
                    message.buttonsMessage ||
                    message.listMessage
                );

                if (requiresPatch) {
                    message = {
                        viewOnceMessage: {
                            message: {
                                messageContextInfo: {
                                    deviceListMetadataVersion: 2,
                                    deviceListMetadata: {},
                                },
                                ...message,
                            },
                        },
                    };
                }
                return message;
            }
})
```

**Comment by makanSukros** [comment]
> Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> 
> ```js
> makeWASocket({
>     patchMessageBeforeSending: (message) => {
>         const requiresPatch = !!(
>             message.buttonsMessage ||
>             message.listMessage
>         );
> 
>         if (requiresPatch) {
>             message = {
>                 viewOnceMessage: {
>                     message: {
>                         messageContextInfo: {
>                             deviceListMetadataVersion: 2,
>                             deviceListMetadata: {},
>                         },
>                         ...message,
>                     },
>                 },
>             };
>         }
>         return message;
>     }
> })
> ```
> 
> *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message

Same result, 

makeWASocket({
            version,
            logger: logger(ev),
            printQRInTerminal: false,
            auth: state,
            syncFullHistory: true,
            getMessage: handlerRetry.messageRetryHandler,
            patchMessageBeforeSending: (message) => {
                const requiresPatch = !!(
                    message.buttonsMessage ||
                    // || message.templateMessage
                    message.listMessage
                );
                if (requiresPatch) {
                    message = {
                        viewOnceMessage: {
                            message: {
                                messageContextInfo: {
                                    deviceListMetadataVersion: 2,
                                    deviceListMetadata: {},
                                },
                                ...message,
                            },
                        },
                    };
                }

                return message;
            }
        })

**Comment by Zedmerlol** [comment]
In my phone I can see the buttons from normal WhatsApp, but I can't see from WhatsApp Business (I'm using the latest version of whatsapp)

**Comment by Phhhp** [comment]
Avistei a novinha no grau saber oq ela quer pau kkkkkkk

{
   "stanzaId": "xxxxxxx",
   "participant": "xxxxxx@s.whatsapp.net",
   "quotedMessage": {
      "buttonsMessage": {
         "contentText": "
┏━━━━❪🎰❫━━━━
┣► 🍊 : 🍋 : 🔔◄┛
┗━━━━❪💰❫━━━━
*Você perdeu...*
",
         "footerText": "Você perdeu...",
         "buttons": [
            {
               "buttonId": "-cassino",
               "buttonText": {
                  "displayText": "Proximo"
               },
               "type": "RESPONSE"
            }
         ],
         "headerType": "EMPTY"
      }
   }
}

**Comment by NowDev** [comment]
@Phhhp ??? KKKKKKKKKKK

**Comment by Phhhp** [comment]
> @Phhhp ??? KKKKKKKKKKK

Todo mundo sério cara tem que ter um gaiato no meio

**Comment by NowDev** [comment]
When receiving a message from a company that uses the oficial WhatsApp API they seems to use other props, this is the format they're using:
![image](https://user-images.githubusercontent.com/33182389/225751410-8e9af175-039b-4391-9706-196aa0b6a644.png)

**Comment by makanSukros** [comment]
> When receiving a message from a company that uses the oficial WhatsApp API they seems to use other props, this is the format they're using: ![image](https://user-images.githubusercontent.com/33182389/225751410-8e9af175-039b-4391-9706-196aa0b6a644.png) `text` changed to `contentText` and types are enums instead of integers, i'm not sure if baileys changes the payload before sending.

@adiwajshing

**Comment by jgcvsoares2012** [comment]
Weird thing is that most of WhatsApp receives without any problem. But a few don’t! We tried here on different scenarios… 6 mobile phones. All WA on last update. 2 of them didn’t receive! Still not working... on WA Web show the message for those 2. On mobile (Android, QA Personal) didn’t receive.

**Comment by makanSukros** [comment]
> Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> 
> ```js
> makeWASocket({
>     patchMessageBeforeSending: (message) => {
>         const requiresPatch = !!(
>             message.buttonsMessage ||
>             message.listMessage
>         );
> 
>         if (requiresPatch) {
>             message = {
>                 viewOnceMessage: {
>                     message: {
>                         messageContextInfo: {
>                             deviceListMetadataVersion: 2,
>                             deviceListMetadata: {},
>                         },
>                         ...message,
>                     },
>                 },
>             };
>         }
>         return message;
>     }
> })
> ```
> 
> *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message
I have tested and listened to incoming and outgoing conversations from whatsapp, it turns out that after using this function, messages sent to whatsapp are correct, there is an additional messageContextInfo variable, but when I look at the log in messages.upsert on the message recipient number button, the additional messageContextInfo variable is not sent, it seems there is filtering on the whatsapp server @adiwajshing @PurpShell

**Comment by makanSukros** [comment]
> Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> 
> ```js
> makeWASocket({
>     patchMessageBeforeSending: (message) => {
>         const requiresPatch = !!(
>             message.buttonsMessage ||
>             message.listMessage
>         );
> 
>         if (requiresPatch) {
>             message = {
>                 viewOnceMessage: {
>                     message: {
>                         messageContextInfo: {
>                             deviceListMetadataVersion: 2,
>                             deviceListMetadata: {},
>                         },
>                         ...message,
>                     },
>                 },
>             };
>         }
>         return message;
>     }
> })
> ```
> 
> *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message

I have tested and listened to incoming and outgoing conversations from whatsapp, it turns out that after using this function, messages sent to whatsapp are correct, there is an additional messageContextInfo variable, but when I look at the log in messages.upsert on the message recipient number button, the additional messageContextInfo variable is not sent, it seems there is filtering on the whatsapp server @adiwajshing @PurpShell

**Comment by makanSukros** [comment]
> > Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> > ```js
> > makeWASocket({
> >     patchMessageBeforeSending: (message) => {
> >         const requiresPatch = !!(
> >             message.buttonsMessage ||
> >             message.listMessage
> >         );
> > 
> >         if (requiresPatch) {
> >             message = {
> >                 viewOnceMessage: {
> >                     message: {
> >                         messageContextInfo: {
> >                             deviceListMetadataVersion: 2,
> >                             deviceListMetadata: {},
> >                         },
> >                         ...message,
> >                     },
> >                 },
> >             };
> >         }
> >         return message;
> >     }
> > })
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message
> 
> I have tested and listened to incoming and outgoing conversations from whatsapp, it turns out that after using this function, messages sent to whatsapp are correct, there is an additional messageContextInfo variable, but when I look at the log in messages.upsert on the message recipient number button, the additional messageContextInfo variable is not sent, it seems there is filtering on the whatsapp server @adiwajshing @PurpShell

![image](https://user-images.githubusercontent.com/72696825/225881032-313345aa-e6c5-4a68-bff0-9307f10e6b21.png)

**Comment by marcelo386** [comment]
+1

**Comment by makanSukros** [comment]
Does anyone have any ideas to fix this problem?

**Comment by manoeuvrelawliet** [comment]
+1

**Comment by Zedmerlol** [comment]
+1

**Comment by arturludovico** [comment]
Botões não funcionam mais após a nova atualização do whatsapp

**Comment by makanSukros** [comment]
It seems that at this time the button and list messages really don't appear on all numbers that have used the latest WhatsApp version

**Comment by hassanfuad96** [comment]
+1

**Comment by deriknyvo** [comment]
something news?

**Comment by fabbioSborges** [comment]
+1

**Comment by dafner** [comment]
Any idea about how to solve this problem?

Thanks

**Comment by BrunoSobrino** [comment]
+1

**Comment by ericdesaoficial** [comment]
+1

**Comment by jcdevsoft** [comment]
+1

**Comment by SarkarJobs** [comment]
+1

**Comment by elrebelde21** [comment]
Otra vez WhatsApp, la volvió a caga con sus actualizaciones xd

**Comment by mmmlves** [comment]
> Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> 
> ```js
> makeWASocket({
>     patchMessageBeforeSending: (message) => {
>         const requiresPatch = !!(
>             message.buttonsMessage ||
>             message.listMessage
>         );
> 
>         if (requiresPatch) {
>             message = {
>                 viewOnceMessage: {
>                     message: {
>                         messageContextInfo: {
>                             deviceListMetadataVersion: 2,
>                             deviceListMetadata: {},
>                         },
>                         ...message,
>                     },
>                 },
>             };
>         }
>         return message;
>     }
> })
> ```
> 
> *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message

Hello!
  I tried the patch, but it's not working. This is a sample of the message I'm sending:
`{
  "image": {
    "url": "https://system.halai.com.br/images/halai-ico.png"
  },
  "caption": "Oi tudo bem? Como podemos ajudá-lo?\n*Se você já em cliente e deseja abrir um chamado, clique em \"Já sou cliente\"*",
  "footer": "",
  "buttons": [
    {
      "buttonId": "output_2",
      "buttonText": {
        "displayText": "Produtos Whatsapp"
      },
      "type": 1
    },
    {
      "buttonId": "output_3",
      "buttonText": {
        "displayText": "Já Sou Cliente"
      },
      "type": 1
    },
    {
      "buttonId": "output_4",
      "buttonText": {
        "displayText": "Outros Produtos"
      },
      "type": 1
    }
  ],
  "headerType": 4
}
`
and the command:
const sentMessage = await sock.sendMessage(jid, msg)

Any suggestion? Thanks for helping!

**Comment by VeLuisS** [comment]
+1

**Comment by rajuaryan21** [comment]
+1

**Comment by mmmlves** [comment]
No solution yet.. 3 days. It looks like buttons and list messages outside the official API is dead .. 🥺

**Comment by tokalink** [comment]
+1

**Comment by PhantomD** [comment]
Same issue.

Below are the latest version that still buttons and list messages working on:
Whatsapp: 2.23.3.79
Whatsapp Business: 2.23.3.77

**Comment by smashah** [comment]
Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:

![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)

[0] is normal message payload 
[1] is a verified business signature
[2] is wa biz api conversation metadata

The big question is can 1 and 2 be spoofed?

When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.

IDK if the smart cookies in Baileys are working on this but if I'm ready to help if you lot need an extra hand.

🙏

**Comment by azudindaem** [comment]
Hiding the message button inside a sandwich like before is quite challenging. it was likely just a temporary workaround.

Currently, WhatsApp only allows the button message to be sent using the official API. I'm not sure about spoofing the WA server, but hopefully we can resolve this issue quickly.

Regardless, we have to assume that the message button dead.

**Comment by Auties00** [comment]
> Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> 
> ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> 
> [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> 
> The big question is can 1 and 2 be spoofed?
> 
> When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> 
> IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> 
> 🙏

I made the same conversations. 
In addition, I also reversed both the standard and business apps for android, but I couldn't find any evidence that those values are being used for validation:
- all the values starting with paid_convo aren't used anywhere
- host_storage, actual_actors and privacy_mode_ts are used but not for validation afaict
- This node is being used 100%: Node[description=hsm, attributes={category=NON_TRANSACTIONAL}]
In the protobuf what's changed is deviceListMetadata, so maybe combing both information will fix it?

**Comment by makanSukros** [comment]
> > Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> > ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> > [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> > The big question is can 1 and 2 be spoofed?
> > When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> > IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> > 🙏
> 
> I made the same observations. In addition, I also reversed both the standard and business apps for android, but I couldn't find any evidence that those values are being used for validation:
> 
> * all the values starting with paid_convo aren't used anywhere
> * host_storage, actual_actors and privacy_mode_ts are used but not for validation afaict
> * This node is being used 100%: Node[description=hsm, attributes={category=NON_TRANSACTIONAL}]
> 
> In the protobuf what's changed is deviceListMetadata, so maybe combing both pieces of information will fix it?
> 
> EDIT: Btw the reason why you couldn't override those parameters is that outgoing messages don't follow the same format as incoming messages. In practice, while Whatsapp sends those values in those nodes for an incoming message, we would need to see how it does it for outgoing messages because it will be 100% different.
> 
> EDIT2: I can confirm that they are not the same, using those will result in bug out the Whatsapp backend causing disconnection

does this have anything to do with the receiptKeyHash value?

**Comment by liimalim** [comment]
> > > Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> > > ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> > > [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> > > The big question is can 1 and 2 be spoofed?
> > > When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> > > IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> > > 🙏
> > 
> > 
> > I made the same observations. In addition, I also reversed both the standard and business apps for android, but I couldn't find any evidence that those values are being used for validation:
> > 
> > * all the values starting with paid_convo aren't used anywhere
> > * host_storage, actual_actors and privacy_mode_ts are used but not for validation afaict
> > * This node is being used 100%: Node[description=hsm, attributes={category=NON_TRANSACTIONAL}]
> > 
> > In the protobuf what's changed is deviceListMetadata, so maybe combing both pieces of information will fix it?
> > EDIT: Btw the reason why you couldn't override those parameters is that outgoing messages don't follow the same format as incoming messages. In practice, while Whatsapp sends those values in those nodes for an incoming message, we would need to see how it does it for outgoing messages because it will be 100% different.
> > EDIT2: I can confirm that they are not the same, using those will result in bug out the Whatsapp backend causing disconnection
> 
> does this have anything to do with the receiptKeyHash value?

I don't think the problem is with receiptKeyHash, because text messages also contain receiptKeyHash and can be displayed on Android

**Comment by azudindaem** [comment]
sender & receiptKeyHash is not randomly generated and we can pass it when sending the message..

**Comment by is99zsq** [comment]
> Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> 
> ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> 
> [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> 
> The big question is can 1 and 2 be spoofed?
> 
> When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> 
> IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> 
> 🙏

[1] is more important i think.  when part [0] is decoded as a business message format, then a verified process will  continue:

block [1] is a protobuf data maybe  like this:

message VerifiedNameCertificate {
    message Details {
        optional uint64 serial = 1;
        optional string issuer = 2;
        optional string verifiedName = 4;
        repeated LocalizedName localizedNames = 8;
        optional uint64 issueTime = 10;
    }
    optional bytes details = 1;
    optional bytes signature = 2;
    optional bytes serverSignature = 3;
}

[detail]
serial:    "verified_name"  from message attribute
issue:    always  "ent:wa"
verifiedNamed:  is your account display name in WABA 
issueTime:  a timestamp 

[signature]
from api-client

[serverSignature]
when server receive the api-client request,   may verify some infos and fill the area

CONCLUSION:  
template message WILL NOT be sent from personal whatsapp number  in  the foreseeable future。

**Comment by Auties00** [comment]
> > Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> > 
> > ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> > 
> > [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> > 
> > The big question is can 1 and 2 be spoofed?
> > 
> > When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> > 
> > IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> > 
> > 🙏
> 
> [1] is more important i think.  when part [0] is decoded as a business message format, then a verified process will  continue:
> 
> block [1] is a protobuf data maybe  like this:
> 
> message VerifiedNameCertificate {
>     message Details {
>         optional uint64 serial = 1;
>         optional string issuer = 2;
>         optional string verifiedName = 4;
>         repeated LocalizedName localizedNames = 8;
>         optional uint64 issueTime = 10;
>     }
>     optional bytes details = 1;
>     optional bytes signature = 2;
>     optional bytes serverSignature = 3;
> }
> 
> [detail]
> serial:    "verified_name"  from message attribute
> issue:    always  "ent:wa"
> verifiedNamed:  is your account display name in WABA 
> issueTime:  a timestamp 
> 
> [signature]
> from api-client
> 
> [serverSignature]
> when server receive the api-client request,   may verify some infos and fill the area
> 
> CONCLUSION:  
> template message WILL NOT be sent from personal whatsapp number  in  the foreseeable future。
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 

As i said this has nothing to do with the issue. 

When you send a message, you are sending a node containing an array of bytes representing the encoded protobuf. When you instead receive a message, you get a node containing a ciphered protobuf. While the incoming and outgoing protobufs remain the same, the wrapper's metadata can change. This is demonstrated, for example, by the "delete for everyone" feature, which requires a flag in the outgoing node that is not there in the incoming one.

As the other user sent a screenshot of the metadata for an incoming message, it has no value for message sending, except that it suggests that the outgoing node contains some different metadata. If you could send a template message from Whatsapp web's client or even from the mobile app, we could inject some code to print out the outgoing nodes and easily find what was changed. The problem here is that, as far as I can tell, these kind of messages can only be sent by the official whatsapp business api, so we don't have access to the outgoing node, but only to the incoming one. In short, we can't fix this issue easily because there is no client with this feature to be reversed.

On the bright side, all of the tokens used for a node's metadata are known from the single byte and double byte list: we could perhaps infer what's needed to fix this issue, but it's not as straight forward. 

Conclusion:
- If the problem is the protobuf, it's something that can be fixed very easily, but this doesn't look to be the case as the properties that are now filled in the device metadata object aren't used for validation in the app as far as I can tell. Furthermore even filling in the correct values, which is something that I have already done(you can do the same by checking out my analyzer to understand where the tokens come from, it's actually very easy), the issue isn't fixed.
- If the problem is the outgoing node's metadata, the issue is far harder to fix as we don't have access to any client that can send a template message(or any other type of button message for that matter) to reverse. Still, this could be fixed as all of the tokens are known and we could infer by their name which parameters need to be included as metadata and their respective values.

**Comment by is99zsq** [comment]
> > > Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> > > ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> > > [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> > > The big question is can 1 and 2 be spoofed?
> > > When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> > > IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> > > 🙏
> > 
> > 
> > [1] is more important i think.  when part [0] is decoded as a business message format, then a verified process will  continue:
> > block [1] is a protobuf data maybe  like this:
> > message VerifiedNameCertificate {
> > message Details {
> > optional uint64 serial = 1;
> > optional string issuer = 2;
> > optional string verifiedName = 4;
> > repeated LocalizedName localizedNames = 8;
> > optional uint64 issueTime = 10;
> > }
> > optional bytes details = 1;
> > optional bytes signature = 2;
> > optional bytes serverSignature = 3;
> > }
> > [detail]
> > serial:    "verified_name"  from message attribute
> > issue:    always  "ent:wa"
> > verifiedNamed:  is your account display name in WABA
> > issueTime:  a timestamp
> > [signature]
> > from api-client
> > [serverSignature]
> > when server receive the api-client request,   may verify some infos and fill the area
> > CONCLUSION:
> > template message WILL NOT be sent from personal whatsapp number  in  the foreseeable future。
> 
> As i said this has nothing to do with the issue.
> 
> When you send a message, you are sending a node containing an array of bytes representing the encoded protobuf. When you instead receive a message, you get a node containing a ciphered protobuf. While the incoming and outgoing protobufs remain the same, the wrapper's metadata can change. This is demonstrated, for example, by the "delete for everyone" feature, which requires a flag in the outgoing node that is not there in the incoming one.
> 
> As the other user sent a screenshot of the metadata for an incoming message, it has no value for message sending, except that it suggests that the outgoing node contains some different metadata. If you could send a template message from Whatsapp web's client or even from the mobile app, we could inject some code to print out the outgoing nodes and easily find what was changed. The problem here is that, as far as I can tell, these kind of messages can only be sent by the official whatsapp business api, so we don't have access to the outgoing node, but only to the incoming one. In short, we can't fix this issue easily because there is no client with this feature to be reversed.
> 
> On the bright side, all of the tokens used for a node's metadata are known from the single byte and double byte list: we could perhaps infer what's needed to fix this issue, but it's not as straight forward.
> 
> Conclusion:
> 
> * If the problem is the protobuf, it's something that can be fixed very easily, but this doesn't look to be the case as the properties that are now filled in the device metadata object aren't used for validation in the app as far as I can tell. Furthermore even filling in the correct values, which is something that I have already done(you can do the same by checking out my analyzer to understand where the tokens come from, it's actually very easy), the issue isn't fixed.
> * If the problem is the outgoing node's metadata, the issue is far harder to fix as we don't have access to any client that can send a template message(or any other type of button message for that matter) to reverse. Still, this could be fixed as all of the tokens are known and we could infer by their name which parameters need to be included as metadata and their respective values.


it is very clear that  the receiver can get the encoded protobuf ( STATUS RECEVIED）and  sucessfully decode， but the message cannot show in the UI (STATUS READ),   something must happend in this step.    so  we focus on the VERIFY feature about the incoming message. 

in a  Api-Client to App  ,   the api-client can generate the correct metadata( signatures) , of course it can pass the verification.

in an  old version ViewOnceMessage Sender to App ,  the WA server didn't see the VERIFY attrs in the msg and let it go through, but the message can be interpreted into TemplateMessage and show out

in a new version  ViewOnceMessage Sender to App ,the WA server also pass through, but the App detect a Non-App feature messages ,  it may recall the VERIFY step,   the message will be dropped when verification is fail.

**Comment by Auties00** [comment]
> > > > Might be pissing into the wind by coming here with this info but this is why buttons stopped triggering notifications on iOS - I can only assume that this is the same reason why it has completely stopped working on android devices. Below is the screenshot of the raw payload sent when a verified wa biz api customer business account sends a paid-for button/template message:
> > > > ![image](https://user-images.githubusercontent.com/5903821/227748414-1e687177-6387-4060-b1f3-b038d154bfc5.png)
> > > > [0] is normal message payload [1] is a verified business signature [2] is wa biz api conversation metadata
> > > > The big question is can 1 and 2 be spoofed?
> > > > When I was testing a possible iOS fix, I made sure the [0] payload was identical. I "hijacked" this biz accounts verfication sig [1] and the convo metadata [2] but it still wouldn't work. My conclusion is [1] and [2] are being checked with wa servers before the mobile app client decides if it wants show you the message - this would effectively make template messages impossible to resurrect.
> > > > IDK if the smart cookies in Baileys are working on this but I'm ready to help if you lot need an extra hand.
> > > > 🙏
> > > 
> > > 
> > > [1] is more important i think.  when part [0] is decoded as a business message format, then a verified process will  continue:
> > > block [1] is a protobuf data maybe  like this:
> > > message VerifiedNameCertificate {
> > > message Details {
> > > optional uint64 serial = 1;
> > > optional string issuer = 2;
> > > optional string verifiedName = 4;
> > > repeated LocalizedName localizedNames = 8;
> > > optional uint64 issueTime = 10;
> > > }
> > > optional bytes details = 1;
> > > optional bytes signature = 2;
> > > optional bytes serverSignature = 3;
> > > }
> > > [detail]
> > > serial:    "verified_name"  from message attribute
> > > issue:    always  "ent:wa"
> > > verifiedNamed:  is your account display name in WABA
> > > issueTime:  a timestamp
> > > [signature]
> > > from api-client
> > > [serverSignature]
> > > when server receive the api-client request,   may verify some infos and fill the area
> > > CONCLUSION:
> > > template message WILL NOT be sent from personal whatsapp number  in  the foreseeable future。
> > 
> > As i said this has nothing to do with the issue.
> > 
> > When you send a message, you are sending a node containing an array of bytes representing the encoded protobuf. When you instead receive a message, you get a node containing a ciphered protobuf. While the incoming and outgoing protobufs remain the same, the wrapper's metadata can change. This is demonstrated, for example, by the "delete for everyone" feature, which requires a flag in the outgoing node that is not there in the incoming one.
> > 
> > As the other user sent a screenshot of the metadata for an incoming message, it has no value for message sending, except that it suggests that the outgoing node contains some different metadata. If you could send a template message from Whatsapp web's client or even from the mobile app, we could inject some code to print out the outgoing nodes and easily find what was changed. The problem here is that, as far as I can tell, these kind of messages can only be sent by the official whatsapp business api, so we don't have access to the outgoing node, but only to the incoming one. In short, we can't fix this issue easily because there is no client with this feature to be reversed.
> > 
> > On the bright side, all of the tokens used for a node's metadata are known from the single byte and double byte list: we could perhaps infer what's needed to fix this issue, but it's not as straight forward.
> > 
> > Conclusion:
> > 
> > * If the problem is the protobuf, it's something that can be fixed very easily, but this doesn't look to be the case as the properties that are now filled in the device metadata object aren't used for validation in the app as far as I can tell. Furthermore even filling in the correct values, which is something that I have already done(you can do the same by checking out my analyzer to understand where the tokens come from, it's actually very easy), the issue isn't fixed.
> > * If the problem is the outgoing node's metadata, the issue is far harder to fix as we don't have access to any client that can send a template message(or any other type of button message for that matter) to reverse. Still, this could be fixed as all of the tokens are known and we could infer by their name which parameters need to be included as metadata and their respective values.
> 
> 
> it is very clear that  the receiver can get the encoded protobuf ( STATUS RECEVIED）and  sucessfully decode， but the message cannot show in the UI (STATUS READ),   something must happend in this step.    so  we focus on the VERIFY feature about the incoming message. 
> 
> in a  Api-Client to App  ,   the api-client can generate the correct metadata( signatures) , of course it can pass the verification.
> 
> in an  old version ViewOnceMessage Sender to App ,  the WA server didn't see the VERIFY attrs in the msg and let it go through, but the message can be interpreted into TemplateMessage and show out
> 
> in a new version  ViewOnceMessage Sender to App ,the WA server also pass through, but the App detect a Non-App feature messages ,  it may recall the VERIFY step,   the message will be dropped when verification is fail.
> 
> 

The problem is that this verification is probably done using the incoming node's metadata, not the protobuf's data. This is also the case for lice location messages for example. Whatsapp sends that metadata based on the outgoing message's metadata that we can't access. That's the problem and the only way to fix it is to what I explained in the second point i brought up in my conclusion.

**Comment by Auties00** [comment]
Perhaps they aren't accepted because there needs to be also this node:
Node.ofAttributes("hsm", Map.of("category", "NON_TRANSACTIONAL")))
Sadly if you add that in your client will be disconnected with a  403 error and the companion will get unpaired.
You'll need to log in again using the SMS for the companion.

**Comment by is99zsq** [comment]
> New information: I added the biz node to the request payload with the following parameters:
> 
> * "host_storage", "1" (HostStorageType proto enum)
> * "actual_actors", "2" (I'm going to guess ActualActorsType proto enum even though it doesn't reach index 2)
> * "privacy_mode_ts", "1612483200" (just a timestamp in seconds)
> 
> The message still gets sent, not displayed but at least it doesn't get rejected.
> 
> Adding any of the paid_convo properties makes the message bounce back even though I observed paid_convo_ts, paid_convo_id and paid_convo_category to be constant for the same chat as I received two subsequent messages from a registered whatsapp business bot with those values staying the same. paid_convo_status instead changed from 1 to 4 and stayed there for subsequent messages (tried from a regular whatsapp web client companion, not business)


here is a full incoming message log from sent from an WABA to APP,  where i can see an empty hsm node  and the biz node, 

<message from='----------------@s.whatsapp.net' type='text' id='BB06B8ECD763BB87EB' verified_name='1489478083652307' verified_level='unknown' notify='-------------' t='1679901069'><enc type='msg' v='2'>MwohBUMziXeqOFLvXwwxsCLdxdPUuqmz8Z1+T72QAdTbVzF4EAAYACLgAT6fDOYSUnt6lOIO6AilmgaJPoa/xdXAX9NPVpyGfu4b99Jwp9PyZ1if4Ap880C8nbE0MwTS15CcycC4KhSu7SRDaRYoPaY+JmL7PNXBgGCUoYwk3KRZM3EcT5WvLNr46hA5edQK2rc4P3q/7PT1s8D937WzY0nBcXBLqOalHgK+nQg8yJoN5r8ZqwT1zaWQyWp/ziqWkw1yyoKmHcmo6kSLv8EZVPCdvuyA8gIAMJXAuCAEi+Pe9rTlz5mnd8grj78EUzbt7TkUAd75YeRf4bcLEj56geRLysKgNuLZditrfoVC/nebkPo=</enc><hsm/><verified_name v='1' verified_level='unknown'>ChwI09XOoMLV0gISBmVudDp3YSIDeGdvUJeKrpYGEkAqBMlJgP2kND5tMDXmDU1PwIPg4GpbntDTqpeaegaJSAh4iIZ+v/lKOxDidA1IXWS/dutAWoauJCRfqHJLCRkDGkDOYCDo1lxkI9t/xfuObQiRUIqOXYNRyE1UVToyuN35cj7SHZmfPS0SEmPjRclBqY1KH7Nqlp17c5baG5EYmUQI</verified_name><biz actual_actors='1' host_storage='2' paid_convo_category='business_initiated' paid_convo_id='a033831d9f193676dd8e89ec59df057e' paid_convo_status='12' paid_convo_ts='1679899560' privacy_mode_ts='1657505049'/></message>
                  I

**Comment by gigantiumweb** [comment]
buttons not work yet

**Comment by smashah** [comment]
@Auties00 im open to going over this on a discord or call or something with other devs (i can be found via my discord in link in openwa main repo).

There's a few theories here but seems like we're coming to the same conclusion - we both may be wrong so.... would be good for all the maintainers to share potential theories in a more real-time format

**Comment by azudindaem** [comment]
I like to see @smashah  and @Auties00  and other developers who are all willing to come together and find solutions for the common good.

**Comment by azudindaem** [comment]
Great! It is still useful even though it does not provide a significant change to allow messages to be displayed on messenger apps. Or maybe its is part of validation.. we are still unclear about this matter.

**Comment by Auties00** [comment]
Managed to fix all button responses(for example ListResponse)
Fixed ListMessage
TemplateMessage still doesn't work
Once I have all of them working I'll create a pr for baileys as well

**Comment by Auties00** [comment]
Fixed ButtonsMessage as well, the only non working is TemplateMessage right now

**Comment by tiagomatrixd** [comment]
> Managed to fix all button responses(for example ListResponse) Fixed ListMessage TemplateMessage still doesn't work Once I have all of them working I'll create a pr for baileys as well

I'm already anxious :)

**Comment by ecatugy** [comment]
> Great! It is still useful even though it does not provide a significant change to allow messages to be displayed on messenger apps. Or maybe its is part of validation.. we are still unclear about this matter.

Looking forward to your pull request. I have several clients waiting, and your contribution will be a great help to the community. Thank you for your assistance!

**Comment by lindionez** [comment]
This brightened my day so much!

**Comment by smashah** [comment]
@Auties00 this is essentially the approach used from the browser based project. the main distinction at the moment is that we're missing keyIndexes for sender and recipient for some reason. Let's see if forcing them into msgctx works.

**Comment by makanSukros** [comment]
> this is essentially the approach used from the browser based project. the main distinction at the moment is that we're missing keyIndexes for sender and recipient for some reason. Let's see if forcing them into msgctx works.

Magic, you're crazy smart dude 👍

**Comment by smashah** [comment]
almost done. Who here has access to a biz api account with template message? or knows of an existng official bot that sends out template message?

**Comment by smashah** [comment]
R.e Key Indexes. It didn't make a difference. What did make a difference was adding payload [2] from here https://github.com/adiwajshing/Baileys/issues/2631#issuecomment-1483955749

The differentiator is getting the correct payload there.

**Comment by NoerOelama** [comment]
> R.e Key Indexes. It didn't make a difference. What did make a difference was adding payload [2] from here [#2631 (comment)](https://github.com/adiwajshing/Baileys/issues/2631#issuecomment-1483955749)
> 
> The differentiator is getting the correct payload there.
> 
> call for help: Please respond here with verified wa biz accounts that send you a hsm/template message when you start a chat with it. Need screenshots as proof and wa link + website page.

I know some numbers. Which part should be taken the screenshot?

**Comment by Auties00** [comment]
> @Auties00 this is essentially the approach used from the browser based project. the main distinction at the moment is that we're missing keyIndexes for sender and recipient for some reason. Let's see if forcing them into msgctx works.

That won't make a difference. I've already fixed the issue for all messages except for templates because they make your device unpaid from its phone number for some reason. I noticed that every time a template is sent in a new chat first an unavailable message comes and then the real one is sent, so maybe that's important as it doesn't happen for any other type of message. Btw if you actually want to see the solution check my repo

**Comment by Auties00** [comment]
> R.e Key Indexes. It didn't make a difference. What did make a difference was adding payload [2] from here https://github.com/adiwajshing/Baileys/issues/2631#issuecomment-1483955749
> 
> The differentiator is getting the correct payload there.
> 
> call for help: Please respond here with verified wa biz accounts that send you a hsm/template message when you start a chat with it. Need screenshots as proof and wa link + website page.

This is also irrelevant, yesterday i noticed that you don't need that node. I'll explain why later as I'm late for school rn

**Comment by smashah** [comment]
All of our implementations broke but for different reasons since our approaches are different, on our end the biz node was the missing piece of the puzzle.

Baileys needs to implement the biz node to get working again. Might not be the whole solution but will get most of the way there.

**Comment by Auties00** [comment]
> All of our implementations broke but for different reasons since our approaches are different, on our end the biz node was the missing piece of the puzzle.
> 
> Baileys needs to implement the biz node to get working again. Might not be the whole solution but will get most of the way there.

As is said you could not implement anything in the device metadata and it would still work. The two things you need are: 
1. Correct mediatype(it's now used in the app to validate button responses)
2. A biz node with no attributes and a particular child that you can find in my repo. Trying to include any of the paid_convo attributes here will make Whatsapp's backend reject the message(aka node timeout). Including the others will only add an (i) button next to each message saying "this business worked with external companies to manage this message". By the way don't worry I'll make a PR for baileys this afternoon so everyone can use it. Templates will come when I find a solution

**Comment by is99zsq** [comment]
> > All of our implementations broke but for different reasons since our approaches are different, on our end the biz node was the missing piece of the puzzle.
> > Baileys needs to implement the biz node to get working again. Might not be the whole solution but will get most of the way there.
> 
> As is said you could not implement anything in the device metadata and it would still work. The two things you need are:
> 
> 1. Correct mediatype(it's now used in the app to validate button responses)
> 2. A biz node with no attributes and a particular child that you can find in my repo. Trying to include any of the paid_convo attributes here will make Whatsapp's backend reject the message(aka node timeout). Including the others will only add an (i) button next to each message saying "this business worked with external companies to manage this message". So no attributes for the biz node is the way
> 
> By the way don't worry I'll make a PR for baileys this afternoon so everyone can use it. Templates will come when I find a solution
> 
> If you want to help it would he more useful if you could try to understand why template buttons get rejected when everything else works. Check my repo and try to figure that out as i haven't been currently. If we manage that, than everything will be back to working


code  viewed and  the final solution is just add a empty biz node with an type node

it can work in list and buttons

and i notices there is a config prop item name "template_hsm"=1  in whatsapp client,  do you think it is an ab test switch and now it is only switch on in template message?

**Comment by Auties00** [comment]
> > > All of our implementations broke but for different reasons since our approaches are different, on our end the biz node was the missing piece of the puzzle.
> > > Baileys needs to implement the biz node to get working again. Might not be the whole solution but will get most of the way there.
> > 
> > As is said you could not implement anything in the device metadata and it would still work. The two things you need are:
> > 
> > 1. Correct mediatype(it's now used in the app to validate button responses)
> > 2. A biz node with no attributes and a particular child that you can find in my repo. Trying to include any of the paid_convo attributes here will make Whatsapp's backend reject the message(aka node timeout). Including the others will only add an (i) button next to each message saying "this business worked with external companies to manage this message". So no attributes for the biz node is the way
> > 
> > By the way don't worry I'll make a PR for baileys this afternoon so everyone can use it. Templates will come when I find a solution
> > 
> > If you want to help it would he more useful if you could try to understand why template buttons get rejected when everything else works. Check my repo and try to figure that out as i haven't been currently. If we manage that, than everything will be back to working
> 
> 
> code  viewed and  the final solution is just add a empty biz node with an type node
> 
> it can work in list and buttons
> 
> and i notices there is a config prop item name "template_hsm"=1  in whatsapp client,  do you think it is an ab test switch and now it is only switch on in template message? 

You have no idea what I had to do to find that out haha 
Also don't forget the mediatype or it won't work 
I don't think that ab prop has nothing to do with it, it simply determines if templates are allowed imo

Observations i made for templates:
1. The hsm node isn't inside the biz node just like all other nodes, it's in a standalone node(just like enc): Node[description=hsm, attributes={category=}] (i can confirm that the category is actually empty, i also tried to see it from Whatsapp's web client thinking it was a parsing error)
2. The enc node has no mediatype(all other buttons have it)
3. When you try to send this payload Whatsapp will unlink your phone number from the app itself and disconnect all of its companions. When you log back in using an SMS the linked devices tab will show an error message saying: "An error occured so all of your devices got disconnected" 

Anyone got any ideas?

**Comment by smashah** [comment]
think this may be the missing peice. When I implement this on template messages the promise hangs and the connection drops. Thankfully no bans tho.

![CleanShot 2023-03-30 at 10 59 00](https://user-images.githubusercontent.com/5903821/228769584-eb9c10d9-0eab-4e20-abd9-ee546faa9550.png)

**Comment by Auties00** [comment]
> think this may be the missing peice. When I implement this on template messages the promise hangs and the connection drops. Thankfully no bans tho.
> 
> ![CleanShot 2023-03-30 at 10 59 00](https://user-images.githubusercontent.com/5903821/228769584-eb9c10d9-0eab-4e20-abd9-ee546faa9550.png)
> 

That is interesting. Try to use those attributes instead of the ones I'm using and see if that fixes it. You can use my repo which has already all of the work done if you haven't implemented it.

**Comment by Auties00** [comment]
Also what if you specify both the sid like that code shows and the empty category like i showed before? Does the connection still get killed?

**Comment by ecatugy** [comment]
> Someone ever answer this problem and i solved by his solutions.. add the property key that name as `patchMessageBeforeSending` in `makeWASocket` function and add this code
> 
> ```js
> makeWASocket({
>     patchMessageBeforeSending: (message) => {
>         const requiresPatch = !!(
>             message.buttonsMessage ||
>             message.listMessage
>         );
> 
>         if (requiresPatch) {
>             message = {
>                 viewOnceMessage: {
>                     message: {
>                         messageContextInfo: {
>                             deviceListMetadataVersion: 2,
>                             deviceListMetadata: {},
>                         },
>                         ...message,
>                     },
>                 },
>             };
>         }
>         return message;
>     }
> })
> ```
> 
> *Edit: if this not work well on your cases, please provide the js code that are you using while trying to send button message

Thank you for the code, but the implementation only works with iOS. On Android, the buttons are still missing.

**Comment by syahransaputra** [comment]
I hope there is a solution for this issue

**Comment by smashah** [comment]
@Auties00 try to add this to the `hsm` node:

`{category: "NON_TRANSACTIONAL"}`

**Comment by Auties00** [comment]
![image](https://user-images.githubusercontent.com/28218457/228889467-961a0a1b-a5e5-48d1-8a79-7e34de811129.png)
This code confirms that the node we need to send looks like this:
Node[description=hsm, attributes={id=TEMPLATE_ID}]
The TEMPLATE_ID can be obtained from the protobuf of the TemplateMessage you want to send.
The problem is that the message still gets rejected, no disconnect though

**Comment by Auties00** [comment]
That code is used when the user reports a message as spam to forward it to whatsapp. 
That's the closest that we can get to knowing how a client would act when it wants to send a template message.
Yet it doesn't work so jokes hahaha

**Comment by syahransaputra** [comment]
it's strange why it appears here wa.me/628118614022

**Comment by Auties00** [comment]
> it's strange why it appears here [wa.me/628118614022](https://wa.me/628118614022)

My conclusion is that the issue that is making template messages not work is completely different.
This code gets the media type of the message(extracted from latest Whatsapp apk):
```
public static String A02(C18750x5 c18750x5, C34711kO c34711kO) {
        int i = c34711kO.A00;
        if ((i & 4) == 4) {
            return "image";
        }
        if ((i & 8) == 8) {
            return "vcard";
        }
        if ((i & 4096) == 4096) {
            return "contact_array";
        }
        if ((i & 16) == 16) {
            return "location";
        }
        if ((i & 65536) == 65536) {
            return "livelocation";
        }
        if (AnonymousClass000.A1N(i & 32, 32)) {
            C29L c29l = c34711kO.A0E;
            C29L c29l2 = c29l;
            if (c29l == null) {
                c29l = C29L.A0P;
            }
            if ((c29l.A01 & 2) == 2) {
                if (c29l2 == null) {
                    c29l2 = C29L.A0P;
                }
                String str = c29l2.A0K;
                return c18750x5.A0E(C32031ep.A01(str)) ? "cataloglink" : c18750x5.A0F(C32031ep.A01(str)) ? "productlink" : "url";
            }
        } else if (c34711kO.A0b()) {
            return "document";
        } else {
            if ((i & 128) == 128) {
                C2BM c2bm = c34711kO.A02;
                if (c2bm == null) {
                    c2bm = C2BM.A0G;
                }
                return c2bm.A0E ? "ptt" : "audio";
            } else if ((i & 256) == 256) {
                AnonymousClass293 anonymousClass293 = c34711kO.A0r;
                if (anonymousClass293 == null) {
                    anonymousClass293 = AnonymousClass293.A0O;
                }
                return anonymousClass293.A0M ? "gif" : "video";
            } else {
                int i2 = c34711kO.A01;
                if ((i2 & 2097152) == 2097152) {
                    return "ptv";
                }
                if ((i & EditorInfoCompat.MEMORY_EFFICIENT_TEXT_LENGTH) != 2048) {
                    if ((i & 2097152) == 2097152) {
                        C38561qs c38561qs = c34711kO.A0n;
                        if (c38561qs == null) {
                            c38561qs = C38561qs.A0I;
                        }
                        return c38561qs.A0H ? "avatar_sticker" : "sticker";
                    } else if ((i & EditorInfoCompat.IME_FLAG_NO_PERSONALIZED_LEARNING) == 16777216) {
                        C28G c28g = c34711kO.A0d;
                        C28G c28g2 = c28g;
                        if (c28g == null) {
                            c28g = C28G.A07;
                        }
                        if ((c28g.A00 & 1) == 1) {
                            return "product";
                        }
                        if (c28g2 == null) {
                            c28g2 = C28G.A07;
                        }
                        if ((c28g2.A00 & 4) == 4) {
                            return "catalog";
                        }
                    } else if ((i & 536870912) == 536870912) {
                        return "order";
                    } else {
                        if ((i & 134217728) == 134217728) {
                            return "list";
                        }
                        if ((i & 1073741824) == 1073741824) {
                            return "list_response";
                        }
                        if ((i2 & 2) == 2) {
                            return "buttons_response";
                        }
                        if ((i2 & 32) == 32) {
                            C32T c32t = c34711kO.A0Q;
                            if (c32t == null) {
                                c32t = C32T.A05;
                            }
                            if (c32t.A01 == 2) {
                                return "native_flow_response";
                            }
                        }
                    }
                }
            }
        }
        return null;
    }
```
Template or hms is not there, so it can't be a mediatype thing making it not work. 
The fix I found for the other messages is not present anywhere in the apk though.
And applying the same logic doesn't work.

**Comment by IanosJunior** [comment]
+1

**Comment by Auties00** [comment]
I've created the draft PR
Haven't been able to figure out templates yet

**Comment by Auties00** [comment]
![js](https://user-images.githubusercontent.com/28218457/228948455-5074f5f2-8325-42bb-9627-7be89a84ceb3.png)
Confirmed that including the hsm node is enough to fix the issue
(a is the biz node, e is the parent node)

**Comment by smashah** [comment]
The missing attributes could probably be extracted from those official wa biz API docker images.

**Comment by Auties00** [comment]
> The missing attributes could probably be extracted from those official wa biz API docker images.

Didn't think about that, let me see

**Comment by Auties00** [comment]
> The missing attributes could probably be extracted from those official wa biz API docker images.

Didn't find anything.
lmk if you do

**Comment by is99zsq** [comment]
> > The missing attributes could probably be extracted from those official wa biz API docker images.
> 
> Didn't find anything. lmk if you do

someone tell me that the business-initial session in WABA ,  the first message to the user must be a TEMPLATE message, so the strict hsm-check workflow in template-message is more understandable, and after the session created, you can send buttons or list or others else.

**Comment by Auties00** [comment]
> > > The missing attributes could probably be extracted from those official wa biz API docker images.
> > 
> > Didn't find anything. lmk if you do
> 
> someone tell me that the business-initial session in WABA ,  the first message to the user must be a TEMPLATE message, so the strict hsm-check workflow in template-message is more understandable, and after the session created, you can send buttons or list or others else.

I need to see the nodes that it sends to fix the issue. Either it's missing metadata or it excepts other nodes to be sent straight after

**Comment by Litodroid** [comment]
+1

**Comment by yuvalwarsh** [comment]
> > > > The missing attributes could probably be extracted from those official wa biz API docker images.
> > > 
> > > 
> > > Didn't find anything. lmk if you do
> > 
> > 
> > someone tell me that the business-initial session in WABA ,  the first message to the user must be a TEMPLATE message, so the strict hsm-check workflow in template-message is more understandable, and after the session created, you can send buttons or list or others else.
> 
> I need to see the nodes that it sends to fix the issue. Either it's missing metadata or it excepts other nodes to be sent straight after


Just a question,
Do you happen to know how can I inject the Baileys session into a browser?

**Comment by lucazin** [comment]
the buttons and list buttons is working folks?

**Comment by faviocabral** [comment]
> it's strange why it appears here [wa.me/628118614022](https://wa.me/628118614022)



> it's strange why it appears here [wa.me/628118614022](https://wa.me/628118614022)

hello friend, this is a API or is a whatsapp web js this bot?

**Comment by alizin-api** [comment]
IT'S THE OFFICIAL API, BUT WHATSAPP WEB JS API, WORKS TOO

**Comment by ecatugy** [comment]
[adiwajshing/Baileys] doesn't work anymore. The buttons haven't been working on WhatsApp Business for a while now, so for those who use it professionally, it's no longer reliable. Customers are left in the lurch. I suggest switching to whatsapp-web.js.

**Comment by Auties00** [comment]
> [adiwajshing/Baileys] doesn't work anymore. The buttons haven't been working on WhatsApp Business for a while now, so for those who use it professionally, it's no longer reliable. Customers are left in the lurch. I suggest switching to whatsapp-web.js.

My guy just proposed switching to a library that uses the same exact approach I found(their pr literally references mine)

**Comment by Jakquellyn** [comment]
not valid in WhatsApp version 2.23.6.76

**Comment by dafner** [comment]
+1

Daniel Fernández
Desarrollo de Software
Tel.: +54.11.5252.4446
Móvil: +54.9.11.5511.1884
________________________________

**Comment by asrofindo** [comment]
not stale

**Comment by TheLinuxMafia** [comment]
> [adiwajshing/Baileys] doesn't work anymore. The buttons haven't been working on WhatsApp Business for a while now, so for those who use it professionally, it's no longer reliable. Customers are left in the lurch. I suggest switching to whatsapp-web.js.

whatsapp-web.js have the same BUG coz it's not a bug, it's blocked by WhatsApp. If u need use it professionally, get Official API and pay any session. In Official API all work done tested by me.

**Comment by ecatugy** [comment]
I'm sorry to hear that Baileys is no longer functioning. If you're looking for an alternative, you may need to explore the official API. It's unfortunate to hear that Baileys is no longer available.

**Comment by TheLinuxMafia** [comment]
> I'm sorry to hear that Baileys is no longer functioning. If you're looking for an alternative, you may need to explore the official API. It's unfortunate to hear that Baileys is no longer available.

I've already done with official API but with Baileys u can have a cheap solution for notify through WA. Form my business i need only text message, for other think i use official API.
